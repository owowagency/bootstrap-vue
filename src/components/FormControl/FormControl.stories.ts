import FormControl from '.';
import {ref} from 'vue';
import {sizes} from '@/composables/useSize';
import {types} from '@/composables/useFormControl';

export default {
    title: 'Forms/Form control',
    component: FormControl,
    argTypes: {
        disabled: {
            control: {type: 'boolean'},
        },
        placeholder: {
            control: {type: 'text'},
        },
        plainText: {
            control: {type: 'boolean'},
        },
        readonly: {
            control: {type: 'boolean'},
        },
        size: {
            control: {type: 'select'},
            options: sizes,
        },
        type: {
            control: {type: 'select'},
            options: types,
        },
    },
};

const template = (args) => ({
    components: {FormControl},
    setup() {
        return {args, value: ref('')};
    },
    template: '<FormControl v-model="value" v-bind="args" />',
});

export const Default = template.bind({});

Default.args = {
    modelValue: 'I am an input',
    placeholder: 'I am a placeholder',
};

export const Autofocused = template.bind({});

Autofocused.args = {
    modelValue: 'I am autofocused',
    autofocus: true,
};

export const Disabled = template.bind({});

Disabled.args = {
    modelValue: 'I am disabled',
    disabled: true,
};

export const PlainText = template.bind({});

PlainText.args = {
    modelValue: 'Plain text',
    plainText: true,
};

export const Readonly = template.bind({});

Readonly.args = {
    modelValue: 'I am readonly',
    readonly: true,
};

export const Type = (args) => ({
    components: {FormControl},
    setup() {
        return {args, types};
    },
    template: '<FormControl v-for="t of types" class="mb-1" :placeholder="t" :type="t" v-bind="args" />',
});
