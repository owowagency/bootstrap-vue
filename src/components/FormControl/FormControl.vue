<template>
    <input
        ref="input"
        :autofocus="autofocus"
        class="form-control"
        :class="classes"
        :disabled="disabled"
        :placeholder="placeholder"
        :readonly="readonly"
        :type="type"
        :value="modelValue"
        @input="emit('update:modelValue', $event.target.value)"
    >
</template>

<script lang="ts">
import {computed, ref} from 'vue';
import {formControlProps} from '@/composables/useFormControl';
import useClasses from '@/composables/useClasses';
import useDisabled from '@/composables/useDisabled';
import useSize from '@/composables/useSize';
</script>

<script lang="ts" setup>
const props = defineProps(formControlProps);

const emit = defineEmits(['update:modelValue']);

const input = ref<HTMLInputElement>();

const {classes} = useClasses(computed(() => [
    useDisabled(props.disabled).disabledClass.value,
    props.plainText ? 'form-control-plaintext' : '',
    props.readonly ? 'readonly' : '',
    useSize(props.size, 'form-control-{0}').sizeClass.value,
]));

defineExpose({input});
</script>
