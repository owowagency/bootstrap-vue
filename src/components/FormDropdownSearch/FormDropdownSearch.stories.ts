import FormDropdownSearch from '.';
import {ref} from 'vue';

export default {
    title: 'Forms/Form dropdown search',
    component: FormDropdownSearch,
    argTypes: {
        disableSearchCache: {
            control: {type: 'boolean'},
        },
    },
};

const defaultArgs = {
    items: [
        {label: 'Item 1'},
        {label: 'Item 2'},
        {label: 'Item 3'},
    ],
};

const template = (args) => ({
    components: {FormDropdownSearch},
    setup: () => ({
        args,
        search: ref(''),
        value: ref(),
    }),
    template: `
        <FormDropdownSearch
            v-model="value"
            v-model:search="search"
            placeholder="Search for item"
            v-bind="args"
        />
    `,
});

export const Default = template.bind({});

Default.args = defaultArgs;
