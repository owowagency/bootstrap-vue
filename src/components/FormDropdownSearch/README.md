# FormDropdownSearch

```vue
<FormDropdownSearch :items="" />
```

## Imported Props

| Prop name  | Description                                  | Type                                              | Values | Default |
| ---------- | -------------------------------------------- | ------------------------------------------------- | ------ | ------- |
| autofocus  | Automatically focus on the input when loaded | boolean                                           | -      | false   |
| disabled   | Disables the component                       | boolean                                           | -      | false   |
| modelValue | The model value                              | string                                            | -      | ''      |
| plainText  | Adds the `form-control-plaintext` class      | boolean                                           | -      | false   |
| readonly   | Makes the component readonly                 | boolean                                           | -      | false   |
| size       | The size of the component                    | [Size](../../composables/useSize) (string)        | -      | 'md'    |
| type       | The type of the component                    | [Type](../../composables/useFormControl) (string) | -      | 'text'  |

## Props

| Prop name   | Description                                             | Type    | Values | Default     |
| ----------- | ------------------------------------------------------- | ------- | ------ | ----------- |
| autoSearch  | Automatically search through the items                  | boolean | -      | true        |
| items       | The items of the dropdown menu                          | Item[]  | -      | () =&gt; [] |
| labelKey    | The key that is used to get the label from a given item | string  | -      | 'label'     |
| modelValue  | The model value                                         | Item    | -      | undefined   |
| placeholder | The placeholder of the form input                       | string  | -      | ''          |
| search      | The current search string                               | string  | -      | ''          |

## Slots

| Name           | Description                  | Bindings        |
| -------------- | ---------------------------- | --------------- |
| dropdownToggle | Displays the dropdown toggle | <br/><br/><br/> |
| noOptions      | _Forwarded to child_         |                 |
| prepend        | _Forwarded to child_         |                 |
| items          | _Forwarded to child_         |                 |
| item           | _Forwarded to child_         |                 |
| append         | _Forwarded to child_         |                 |

## Events

| Event name        | Properties                                 | Description                            |
| ----------------- | ------------------------------------------ | -------------------------------------- |
| update:modelValue | **value** `any` - The new value            | Fired when the model value is updated  |
| update:search     | **search** `string` - The new search value | Fired when the search value is updated |
