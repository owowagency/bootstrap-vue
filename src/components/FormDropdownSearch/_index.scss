.form-dropdown-search {
    input {
        cursor: pointer;
    }
}
