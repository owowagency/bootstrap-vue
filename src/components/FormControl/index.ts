import FormControl from './FormControl.vue';

export default FormControl;
