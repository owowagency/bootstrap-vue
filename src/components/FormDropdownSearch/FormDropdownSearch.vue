<template>
    <FormDropdown
        ref="formDropdown"
        v-model="modelValue"
        :items="filteredItems"
        :label-key="labelKey"
    >
        <template #dropdownToggle>
            <slot name="dropdownToggle">
                <FormControl
                    ref="formControl"
                    v-model="searchValueDisplayed"
                    class="form-select"
                    data-bs-toggle="dropdown"
                    :placeholder="searchValueCached || placeholder"
                    v-bind="$attrs"
                    @focus="showMenu"
                />
            </slot>
        </template>

        <!-- Rollup does not dynamically overriding child slots. See issue #35
        <template
            v-for="slotName in formDropdownSlots"
            #[slotName]="slotScope"
        >
            <slot
                :name="slotName"
                v-bind="slotScope"
            />
        </template> -->

        <template #noOptions="slotScope">
            <slot
                name="noOptions"
                v-bind="slotScope"
            />
        </template>

        <template #prepend="slotScope">
            <slot
                name="prepend"
                v-bind="slotScope"
            />
        </template>

        <template #items="slotScope">
            <slot
                name="items"
                v-bind="slotScope"
            />
        </template>

        <template #item="slotScope">
            <slot
                name="item"
                v-bind="slotScope"
            />
        </template>

        <template #append="slotScope">
            <slot
                name="append"
                v-bind="slotScope"
            />
        </template>
    </FormDropdown>
</template>

<script lang="ts" setup>
import {ComponentPublicInstance, PropType, computed, onMounted, ref, watch} from 'vue';
import FormControl from '@/components/FormControl';
import FormDropdown from '@/components/FormDropdown';
import {Item} from '@/composables/useFormSelect';
import useBootstrapInstance from '@/composables/useBootstrapInstance';

const props = defineProps({
    autoSearch: {
        type: Boolean,
        default: true,
    },
    items: {
        type: Array as PropType<Item[]>,
        default: () => [],
    },
    labelKey: {
        type: String,
        default: 'label',
    },
    modelValue: {
        type: Object as PropType<Item>,
        default: undefined,
    },
    placeholder: {
        type: String,
        default: '',
    },
    search: {
        type: String,
        default: '',
    },
});

const emit = defineEmits(['update:modelValue', 'update:search']);

const filteredItems = computed(() => {
    if (searchValue.value === '' || !props.autoSearch) {
        return props.items;
    }

    return props.items
        .filter(i =>
            i.label
                .toLowerCase()
                .includes(searchValue.value.toLocaleLowerCase()),
        );
});

const formControl = ref<ComponentPublicInstance<typeof FormControl>>();

const formDropdown = ref<ComponentPublicInstance<typeof FormDropdown>>();

const {bsInstance: bsDropdown} = useBootstrapInstance(
    'Dropdown',
    formControl,
);

const modelValue = computed({
    get: () => props.modelValue,
    set: v => emit('update:modelValue', v),
});

const modelValueLabel = computed(() => modelValue.value?.[props.labelKey]);

const searchValue = ref<string>(modelValueLabel.value || props.search);

/**
 * Is used to display a temporary value in the input while still emitting an
 * empty `search` to the parent. This will for example happen when this
 * component has a `modelValue` and the user focuses on the input. The parent
 * should receive an empty `search` so it can display more options than options
 * that match the `labelKey` of the `modelValue`. It acts af is there is no
 * `search` while still displaying the `modelValue`.
 */
const searchValueCached = ref<string>('');

const searchValueDisplayed = computed({
    get: () => {
        return searchValueCached.value !== ''
            ? ''
            : searchValue.value;
    },
    set: s => {
        searchValue.value = s;

        emit('update:search', s);
    },
});

onMounted(() => {
    // Get the dropdown toggle element manually because it is possible to
    // override it with the `dropdownToggle` slot.
    const dropdownToggle = formDropdown.value
        .$el
        .querySelector('[data-bs-toggle="dropdown"]');

    dropdownToggle?.addEventListener('shown.bs.dropdown', () => {
        (dropdownToggle.input as HTMLInputElement)?.focus();

        if (modelValueLabel.value) {
            searchValueCached.value = modelValueLabel.value;
        }

        searchValue.value = '';

        if (props.search !== '') {
            emit('update:search', '');
        }
    });

    dropdownToggle?.addEventListener('hidden.bs.dropdown', () => {
        (dropdownToggle.input as HTMLInputElement)?.blur();

        if (!modelValue.value) {
            searchValue.value = '';
        } else {
            searchValue.value = modelValueLabel.value;
        }

        searchValueCached.value = '';
    });
});


const showMenu = () => {
    if (bsDropdown.value) {
        // TODO: BS calls `toggle` once a user clicks on any
        // `data-bs-toggle="dropdown"`. This event is being fired after the
        // `focus` event, thus will result in hiding the menu after we call
        // `show`. `await nextTick()` also does not solve this problem.
        setTimeout(() => bsDropdown.value.show(), 200);
    }
};

watch(
    modelValue,
    newValue => {
        if (newValue) {
            searchValue.value = newValue[props.labelKey];
        } else {
            searchValue.value = '';
        }
    },
);

watch(
    () => props.search,
    s => searchValue.value = s,
);

// Rollup does not like dynamically overriding slots so this is not used for now.
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const formDropdownSlots = [
    'noOptions',
    'prepend',
    'items',
    'item',
    'append',
];

defineExpose({bsDropdown, formControl});
</script>
