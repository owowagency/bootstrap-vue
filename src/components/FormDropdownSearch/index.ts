import FormDropdownSearch from './FormDropdownSearch.vue';

export default FormDropdownSearch;
