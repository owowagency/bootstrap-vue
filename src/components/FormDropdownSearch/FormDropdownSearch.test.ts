import {mount, shallowMount} from '@vue/test-utils';
import {nextTick, ref} from 'vue';
import FormControl from '@/components/FormControl';
import FormDropdownSearch from '.';

const items = [
    {label: 'Item 1'},
    {label: 'Item 2'},
    {label: 'Item 3'},
];

describe('template', () => {
    componentRenderTest(FormDropdownSearch, {props: {items}}, false);

    componentSlotRenderTest(FormDropdownSearch, 'dropdownToggle', {shallow: false});

    componentSlotRenderTest(FormDropdownSearch, 'noOptions', {shallow: false});

    componentSlotRenderTest(FormDropdownSearch, 'prepend', {shallow: false});

    componentSlotRenderTest(FormDropdownSearch, 'items', {shallow: false});

    componentSlotRenderTest(FormDropdownSearch, 'item', {
        props: {items},
        shallow: false,
    });

    componentSlotRenderTest(FormDropdownSearch, 'append', {shallow: false});

    it('calls showMenu when focussing on FormControl', () => {
        const wrapper = mount(FormDropdownSearch, {
            global: {
                mocks: {
                    showMenu: jest.fn(),
                },
            },
        });

        wrapper.findComponent(FormControl).trigger('focus');

        expect(wrapper.vm.showMenu).toBeCalled();
    });
});

describe('filteredItems', () => {
    it('filters items when it has a searchValue and autoSearch is enabled', () => {
        const wrapper = shallowMount(FormDropdownSearch, {props: {items}});

        wrapper.vm.searchValue = '1';

        expect(wrapper.vm.filteredItems).toEqual([items[0]]);
    });

    it('does not filter items when it does not have a searchValue and autoSearch is enabled', () => {
        const wrapper = shallowMount(FormDropdownSearch, {props: {items}});

        wrapper.vm.searchValue = '';

        expect(wrapper.vm.filteredItems).toEqual(items);
    });

    it('does not filter items when it has a searchValue and autoSearch is disabled', () => {
        const wrapper = shallowMount(
            FormDropdownSearch,
            {
                props: {
                    autoSearch: false,
                    items,
                },
            },
        );

        wrapper.vm.searchValue = '1';

        expect(wrapper.vm.filteredItems).toEqual(items);
    });
});

describe('modelValue', () => {
    it('sets searchValue to its label', async() => {
        const modelValue = ref();

        const wrapper = shallowMount(FormDropdownSearch, {props: {modelValue}});

        modelValue.value = items[0];

        await nextTick();

        expect(wrapper.vm.searchValue).toBe(items[0].label);
    });

    it('resets searchValue if undefined', async() => {
        const modelValue = ref(items[0]);

        const wrapper = shallowMount(FormDropdownSearch, {props: {
            modelValue,
        }});

        expect(wrapper.vm.searchValue).toBe(items[0].label);

        modelValue.value = undefined;

        await nextTick();

        expect(wrapper.vm.searchValue).toBe('');
    });
});

describe('search', () => {
    it('sets searchValue on change', async() => {
        const search = ref();

        const wrapper = shallowMount(FormDropdownSearch, {props: {search}});

        search.value = 'changed';

        await nextTick();

        expect(wrapper.vm.searchValue).toBe(search.value);
    });
});

describe('searchValue', () => {
    it('is equal to modelValueLabel by default', async() => {
        const modelValue = ref(items[0]);

        const wrapper = shallowMount(FormDropdownSearch, {props: {modelValue}});

        expect(wrapper.vm.searchValue).toBe(items[0].label);
    });

    it('is equal to search by default when there is no modelValueLabel', async() => {
        const modelValue = ref();

        const wrapper = shallowMount(FormDropdownSearch, {props: {
            modelValue,
            search: 'test',
        }});

        expect(wrapper.vm.searchValue).toBe('test');
    });
});

describe('searchValueDisplayed', () => {
    it('is empty when searchValueCached has a value', () => {
        const wrapper = shallowMount(FormDropdownSearch);

        wrapper.vm.searchValueCached = 'I am not empty';

        wrapper.vm.searchValue = 'do not display me';

        expect(wrapper.vm.searchValueDisplayed).toEqual('');
    });

    it('is equal to searchValue when searchValueCached does not have a value', () => {
        const wrapper = shallowMount(FormDropdownSearch);

        wrapper.vm.searchValueCached = '';

        wrapper.vm.searchValue = 'display me';

        expect(wrapper.vm.searchValueDisplayed).toEqual('display me');
    });

    it('sets search value and emits', () => {
        const wrapper = shallowMount(FormDropdownSearch);

        wrapper.vm.searchValueDisplayed = 'new search value';

        expect(wrapper.vm.searchValue).toEqual('new search value');

        expect(wrapper.emitted('update:search')).toEqual([['new search value']]);
    });
});

describe('onMounted', () => {
    describe('listening to shown.bs.dropdown', () => {
        it('focusses on dropdownToggle', () => {
            // TODO: currently unable to create a test for this.
        });

        it('sets searchValueCached to modelValueLabel', () => {
            const modelValue = ref(items[0]);

            const wrapper = mount(FormDropdownSearch, {props: {modelValue}});

            const event = new Event('shown.bs.dropdown');

            wrapper.find('[data-bs-toggle="dropdown"]').element.dispatchEvent(event);

            expect(wrapper.vm.searchValueCached).toBe(items[0].label);
        });

        it('does not set searchValueCached to modelValueLabel when empty', () => {
            const wrapper = mount(FormDropdownSearch);

            wrapper.vm.searchValueCached = 'do not change me';

            const event = new Event('shown.bs.dropdown');

            wrapper.find('[data-bs-toggle="dropdown"]').element.dispatchEvent(event);

            expect(wrapper.vm.searchValueCached).toBe('do not change me');
        });

        it('empties searchValue', () => {
            const wrapper = mount(FormDropdownSearch);

            wrapper.vm.searchValue = 'not empty';

            const event = new Event('shown.bs.dropdown');

            wrapper.find('[data-bs-toggle="dropdown"]').element.dispatchEvent(event);

            expect(wrapper.vm.searchValue).toBe('');
        });

        it('emits searchValue when search is not empty', () => {
            const wrapper = mount(FormDropdownSearch, {props: {search: 'not empty'}});

            const event = new Event('shown.bs.dropdown');

            wrapper.find('[data-bs-toggle="dropdown"]').element.dispatchEvent(event);

            expect(wrapper.emitted('update:search')).toEqual([['']]);
        });

        it('does not emit searchValue when search is empty', () => {
            const wrapper = mount(FormDropdownSearch, {props: {search: ''}});

            const event = new Event('shown.bs.dropdown');

            wrapper.find('[data-bs-toggle="dropdown"]').element.dispatchEvent(event);

            expect(wrapper.emitted('update:search')).toBeUndefined();
        });
    });

    describe('listening to hidden.bs.dropdown', () => {
        it('blurs on dropdownToggle', () => {
            // TODO: currently unable to create a test for this.
        });

        it('empties searchValue when there is no modelValue', () => {
            const wrapper = mount(FormDropdownSearch);

            wrapper.vm.searchValue = 'not empty';

            const event = new Event('hidden.bs.dropdown');

            wrapper.find('[data-bs-toggle="dropdown"]').element.dispatchEvent(event);

            expect(wrapper.vm.searchValue).toBe('');
        });

        it('sets searchValue to modelValueLabel', () => {
            const modelValue = ref(items[0]);

            const wrapper = mount(FormDropdownSearch, {props: {modelValue}});

            wrapper.vm.searchValue = 'something';

            expect(wrapper.vm.searchValue).toBe('something');

            const event = new Event('hidden.bs.dropdown');

            wrapper.find('[data-bs-toggle="dropdown"]').element.dispatchEvent(event);

            expect(wrapper.vm.searchValue).toBe(items[0].label);
        });

        it('empties searchValueCached', () => {
            const wrapper = mount(FormDropdownSearch);

            wrapper.vm.searchValueCached = 'not empty';

            const event = new Event('hidden.bs.dropdown');

            wrapper.find('[data-bs-toggle="dropdown"]').element.dispatchEvent(event);

            expect(wrapper.vm.searchValueCached).toBe('');
        });
    });
});
