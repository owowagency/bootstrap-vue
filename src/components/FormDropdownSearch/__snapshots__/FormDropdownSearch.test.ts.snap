// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template renders default 1`] = `
"<div class=\\"dropdown form-dropdown\\"><input class=\\"form-control form-control-md form-select\\" placeholder=\\"\\" type=\\"text\\" data-bs-toggle=\\"dropdown\\">
  <ul class=\\"dropdown-menu w-100\\">
    <li><a class=\\"dropdown-item\\" href=\\"#\\">Item 1</a></li>
    <li><a class=\\"dropdown-item\\" href=\\"#\\">Item 2</a></li>
    <li><a class=\\"dropdown-item\\" href=\\"#\\">Item 3</a></li>
  </ul>
</div>"
`;
